README
ctronome v0.5.5

GNU/GPL copying rules apply, see COPYING

DISCLAIMER
   THIS SOFTWARE COMES AS-IS. ALTHOUGH THE AUTHOR DOES ITS BEST TO MAKE
   THIS PROGRAM SAFE TO USE, MISTAKE HAPPENS. USE IT AT YOUR OWN RISK.

   THE AUTHOR DOES NOT TAKE ANY RESPONSIBILITY FOR WHAT THIS SOFTWARE
   DOES TO YOUR PC. BACKUP FREQUENTLY.

INTRODUCTION
   Ctronome is a very-very simple programmable metronome for Linux console,
   with /dev/dsp (OSS) audio output. BPM ranges from 30 to 250, tacts like 3/4, 7/8
   etc. are possible. It plays external WAV files, so one could use custom
   sound samples.
  
   Aim is to keep it as simple as it can be so dummies like me may understand
   and modify it. (The source code looks ugly though;)

   Non-OSS sound systems may need additonal utils/libraries to 
   provide OSS compatibility.

COMPILE-TIME CONFIGURATION:
   edit ctronome.h for defaults, the most important ones are:
   metronomewav1 = "/usr/share/ctronome/metronome1.wav\0";
   metronomewav2 = "/usr/share/ctronome/metronome2.wav\0";
   dspdev = "/dev/dsp\0";

   They speak for themselves.
 
COMPILATION:
   Only the very basic C libraries are needed that I believe to exist
   on virtally all Linuxes. You need to have Linux kernel header
   files installed (they come with the Linux kernel source), and OSS
   (or OSS compatible) soundcard support.

   Go into ctronome source directory and type 'make'.
   It should compile whithout errors.
  
INSTALLATION:
   Ctronome is a single binary. You can put it anywhere you want in
   your local or remote filesystem. You need two, preferably short
   WAV files as metronome sample, put in the place you specified in
   ctronome.h, or specified on the command line.

   There are two one-channel and two two-channel (both channels are the same)
   wav samples included, recorded from a real metronome. You have to install the one that
   fits your sound system (see NOTES).

RUNNING:
 ctronome -b <bpm> -t <bpt> -c <count> -w1 <wav1> -w2 <wav2> -d <dsp_device> \
          -p <program> -h -v
    -b <bpm>            beat per minute default: 60/4, range: 30-250/1-20
    -t <bpt>            beat per tact default: 1/4, range: 1-50/1-50
    -p <filename>       program file
    -c <count>          play tact/program <count> times then exit
    -w1 <filename>      wav to use for first beat of tact
                         default: /usr/share/ctronome/metronome1.wav
    -w2 <filename>      wav to use for other beat of tact
                         default: /usr/share/ctronome/metronome2.wav
    -d <device>         dsp device, default /dev/dsp
    -h                  display help screen
    -debug              display debug messages

   See prog_example.txt to learn how to program ctronome.

   For BPM you can specify the base note. E.g. 60/4 means:
   60 1/4 notes per minute -> length of 1/4 note is 1 second.

   For BPT you can also define the base note. E.g. 4/4 means:
   4 1/4 notes per tact -> length of a tact is 4 seconds.

   As you can see, it is now possible to specify the BPM in 1/4 notes,
   and then you can use BPT like 7/8, ctronome calculates lenght of 1/8 note
   (which is half of the 1/4 in this example) and plays seven of them for a
   tact.

   If the base note for BPM is not specified, it defaults to the base note
   of the tact.

   It requires two 16 bit per sample WAV format, mono or stereo, and preferably
   at least 44100Hz sample rate. *The two WAV formats must be the same!*
   Please note that if your sound system does not support the exact format
   provided by the WAV files, it will try its best to play it anyway, but
   the result will be unpredictable. Debug mode will tell you the details.

RUNNING ON SYSTEMS WITH PulseAudio:
   On modern Linux distributions like Ubuntu 12.10 with PulseAudio, you may
   experience the following error when running ctronome:
	FATAL: cannot open dsp device /dev/dsp
	/dev/dsp: No such file or directory		

   The solution is to either install OSS compatibility modules or install
   the pulsaudio-utils package that contains a wrapper program called padsp.
   Use this wrapper to run ctronome:

	padsp ctronome <ctronome options>

   Or you could use a simple shell script:
	#!/bin/sh
	padsp ctronome $@

RUNNING ON SYSTEMS WITHOUT OSS SUPPORT
   As far as I know, Alsa and PulseAudio both have means to emulate OSS
   either by installing compatibility modules or via a wrapper program.
   Please refer to the documentation of the respective audio system.

   If you have a steb-by-step guide for a specific configuration, please
   send it in and I will include it in the ctronome documentation.

SEND BUGREPORTS TO:
   ctronome at kign dot org

CREDITS:
   dsp code is based on gTick by Alex Roberts <bse@dial.pipex.com>,
            and also the OSS Programmers Guide

   Patch to eliminate warnings with Gcc4, and included WAV samples digitally
   remastered to be more pronounced and precise by Martin Tarenskeen.

   Patch to fix amd64 platform bug by Dmitry Kazarov

   and help from my friends: Zola & Arpy.



ChangeLog
v0.5.5
   - Documentation update: running ctronome with pulseaudio (modern Ubuntu distributions etc)
     and some minor documentation touch-ups.
     NO change to the ctronome code, so no need to recompile and reinstall.

v0.5.4
   - fixed amd64 platform bug that caused ctronome to produce no sound on amd64 platform
     due to incorrect width of integers used in bitwise operations, thanks to Dmitry Kazarov
     http://www.freebsd.org/cgi/query-pr.cgi?pr=ports/170207

v0.5.3
   - added #include <stdlib.h> to routines.c also. Silly me, forgot to
     do that. Again, thanks to Martin for letting me know.

v0.5.2
   - included WAV samples digitally remastered by Martin Tarenskeen
     to be more pronounced and precise.
   - added #include <stdlib.h> to ctronome.c to eliminate compile-time
     warnings with newer (>4) GCC versions

v0.5.1
   - new parameter processing bug introduced in v0.5 fixed
   - some old bugs are fixed around parameter and program processing

v0.5
   - WAV format is detected, and DSP is set up accordingly
   - included 2 channel (stereo) wav format for those sound
     systems that do not support mono mode (see NOTES/README)
   - bugfix in parameter processing that caused bpm base note
     to be ignored

v0.4
   - dsp.c code cleanup, according to OSS Programmers Guide.
   - implemented some debug messages, mainly for DSP procedures

v0.3
   - fixed dsp pattern length setup that
     caused silence at certain bpm rates

v0.2
   - fixed possible segfault bug in argument processing 

   + it is now programmable
   + more configurable bpm and bpt parameters
      allowing to specify the base note for BPM and beat per tact
   + number of repetition can be specified via command-line

v0.1 first version. supports:
     30 < beat per minute < 250
     1 < beat per tact < 50 
     44100Hz 16bit mono WAV sound via /dev/dsp



dsp.c
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <stdlib.h>
#include <linux/soundcard.h>
#include "ctronome.h"

static int dsp_8bit_unsigned_format = AFMT_U8; /* 0x00000008 unsigned 8 bit*/
static int dsp_16bit_signed_format = AFMT_S16_LE; /* 0x00000010 signed 16 bit little endianess*/

static int debug_dsp_speed, debug_dsp_channels, debug_dsp_depth, supported_dsp_formats;

int dsp_init(byte *dsp_dev,word bitspersample, word channels, dword samplerate){
 int dsp_handle;

 dsp_format = 0;
 if (debug) printf ("debug: dsp.c: bitspersample: >%d<\n",bitspersample);
 /* check parameters */
 if (bitspersample == 8)
  dsp_format = dsp_8bit_unsigned_format;
 if (bitspersample == 16){
  dsp_format = dsp_16bit_signed_format;
 }

 if(!(dsp_format)){
  printf ("FATAL: only 8 and 16 bits per sample are supported. (got: %d\n",bitspersample);
  exit (1);
 }
 if (debug) printf ("debug: dsp.c: wanted AFMT: >%d<\n",dsp_format);

 if ((channels < 1) && (channels > 2)){
  printf ("FATAL: number of channels must be 1 (mono) or 2 (stereo) (got: %d\n",channels);
  exit (1);
 }
 dsp_channels = channels;

 if ((samplerate < 8000) && (samplerate > 96000)){
  printf ("FATAL: samplerate must be between 8000 and 96000 (got: %d\n",samplerate);
  exit (1);
 }
 dsp_speed = samplerate;

 /* Initialise dsp_dev */
 if (debug) printf ("debug: opening dsp '%s'\n",dsp_dev);
 if ((dsp_handle = open (dsp_dev, O_WRONLY)) == -1){
  printf ("FATAL: cannot open dsp device %s\n",dsp_dev);
  perror(dsp_dev);
  exit (1);
 }
 if (debug) printf ("debug: dsp opened successfully\n",dsp_dev);

 if (debug){
  if (ioctl(dsp_handle, SNDCTL_DSP_GETFMTS, &supported_dsp_formats) == -1){
   printf ("FATAL: unable to get supported formats for %s\n",dsp_dev);
   perror("SNDCTL_DSP_GETFMT");
   exit(1);  
  }
  printf("debug: supported dsp formats: '%d'\n",supported_dsp_formats);
 }

 if (debug) printf ("debug: dsp: set format: SNDCTL_DSP_SETFMT(%d)\n",dsp_format);
 if ((ioctl (dsp_handle, SNDCTL_DSP_SETFMT, &dsp_format)) == -1){
  printf ("FATAL: unable to set output format for %s\n",dsp_dev);
  perror("SNDCTL_DSP_SETFMT");
  exit(1);  
 }

 if (debug) printf ("debug: dsp: set format returned '%d'\n",dsp_format);
 if ((bitspersample == 8) && (dsp_format != dsp_8bit_unsigned_format)){
  printf("FATAL: your dsp device does not seem to support unsigned 8 bit (AFMT_8U) format\n");
  exit(1);
 }

 if ((bitspersample == 16) && (dsp_format != dsp_16bit_signed_format)){
  printf("FATAL: your dsp device does not seem to support signed 16 bit (AFMT_S16_LE) format\n");
  exit(1);
 }

 /* Set dsp channels */
 if (debug) printf ("debug: dsp: set up channels: SNDCTL_DSP_CHANNELS(%d)\n",dsp_channels);
 /* ioctl (dsp_handle, SNDCTL_DSP_CHANNELS, &dsp_channels); */
 if (ioctl (dsp_handle, SNDCTL_DSP_CHANNELS, &dsp_channels) == -1){
  printf ("FATAL: unable to set no. of channels for %s\n",dsp_dev);
  perror("SNDCTL_DSP_CHANNELS");
  exit(1);
 }

 if (debug) printf ("debug: dsp: set no. of channels returned '%d'\n",dsp_channels);
 if (dsp_channels != channels){
  printf ("FATAL: your wav has %d channels, while your DSP only supports %d channels.\n",channels,dsp_channels);
  exit(1);
 }

 /* Set the DSP rate (in Hz) */
 if (debug) printf ("debug: dsp: set up speed (Hz): SNDCTL_DSP_SPEED(%d)\n",dsp_speed);
 if (ioctl (dsp_handle, SNDCTL_DSP_SPEED, &dsp_speed) == -1){
  printf ("FATAL: unable to set DSP speed for %s\n",dsp_dev);
  perror("SNDCTL_DSP_SPEED");
  exit(1);
 }

 if (debug) printf ("debug: dsp: set speed returned '%d'\n",dsp_speed);
 if (debug && (dsp_speed != samplerate)){
  printf ("debug: dsp: wav samplerate and dsp samplerate differs, will sound funny\n");
 }

 return(dsp_handle);
}

void dsp_close (byte dsp_handle)
{
 close(dsp_handle);
}

void dsp_write (byte dsp_handle,byte *from,dword count){
 dword bytes_written;
 bytes_written = write (dsp_handle,from,count);
 if (bytes_written < count){
  printf ("FATAL: tried to write %d bytes to dsp device,\n\
only %d is written\n",count,bytes_written);
  exit(1);  
 }
 return;
}



ctronome.h
#include <sys/soundcard.h>
#include <stdint.h>

#define MYNAME "ctronome"
#define VERSION "0.5.4"
#define CREDITS "homepage: http://ctronome.kign.org/\n"

/* set up these three variables to your system */
static char *metronomewav1 = "/usr/share/ctronome/metronome1.wav\0";
static char *metronomewav2 = "/usr/share/ctronome/metronome2.wav\0";
static char *dspdev = "/dev/dsp\0";

/* means: 1/4, 1/8 etc. */
static int default_base_note = 4;

char *programfile;

#define HELP "usage: ctronome <parameters>\n\
               valid parameters are:\n\
                -b <bpm>            beat per minute\n\
                -t <bpt>            beat per tact\n\
                -p <filename>       program file\n\
                -c <count>          play tact/program <count> times then exit\n\
                -w1 <filename>      wav to use for first beat of tact\n\
                -w2 <filename>      wav to use for other beat of tact\n\
                -d <device>         dsp device\n\
                -h                  display this help screen\n\
                -v                  print version\n\
           e.g: ctronome -b 60 -t 4\n\
           for defaults/limits and required WAV format see README\n"

/* my lazy type definitions */
typedef uint32_t DWORD;
typedef uint16_t WORD;
typedef uint8_t BYTE;
typedef DWORD dword;
typedef WORD word;
typedef BYTE byte;

word dsp_channels;
dword dsp_speed;
word dsp_format;
byte dsp_depth;

static int count = 1; /* tact counter */
static int pcount = 1; /* repeat tact/program pcount times then exit 0 = endless */
static int pdecrease = 0;

static int bpm[2] = {60,4}; /* 60 bpm is given for 1/4 notes */
static int bpt[2] = {1,4}; /* beat per tact */
static byte slash = 47; /* the / character */
static byte hashmark = 35; /* the # character */
static byte space = 32; /* the   character */

int dsp_init(byte *,word,word,dword);
void next_program(FILE *); /* process the next line of program */
void parm_init(int, char *[]);
void dsp_close(byte);
void dsp_write(byte, byte *, dword);

byte debug;


routines.c
#include <stdio.h>
#include <syslog.h>
#include <unistd.h>
#include "ctronome.h"
#include "routines.h"

dword str_dcopy(byte *source, byte *destination, byte delimiter, dword max)
{
 dword counter;
 counter = 0;
 while ( (source[counter] != 0) && (source[counter] != delimiter) &&
         (counter < max) ) destination[counter] = source[counter++];

 if (source[counter] != delimiter)
  return(0);
 destination[counter] = 0;
 return(counter);
}

dword str_copy(byte *source, byte *destination, dword max)
{
 dword counter;
 counter = 0;
 while ( (source[counter] != 0) &&
         (counter < max) ) destination[counter] = source[counter++];
 destination[counter] = 0;
 return(counter);
}

byte str_compare(byte *a, byte *b, byte delimiter)
{
 dword i = 0;
 while( (a[i] == b[i]) &&
        (a[i] != 0) ) i++;
 if( (a[i] == 0) &&
     ( (b[i] == 0) || (b[i] == delimiter) ) )
  return(1);
 return(0);
}

int str_search(byte *buffer,byte search)
{
 dword count = 0;
 dword i = 0;
 while((buffer[i] != 0) && (buffer[i++] != search));
 if (buffer[--i] == search) return(i);
 return(-1);
}

dword str_replace(byte *buffer,byte search,byte replace)
{
 dword count = 0;
 dword i = 0;
 while(buffer[i] != 0){
  if (buffer[i] == search){
   buffer[i] = replace;
   count++; 
  }
  i++;
 }
 return(count);
}

dword getnextline(byte *buffer,FILE *filehandle, dword max)
{
 dword length;
 length = 0;
 while(fread(&buffer[length],1,1,filehandle) && 
       (buffer[length++] != '\n') &&
       (length<max)); 

 if (length == max){
    printf("line buffer overflow\n");
    return(0); 
 } 
 buffer[length] = 0; 
 return(length);
}

FILE *openread(byte *filename)
{
 FILE *filehandle;
 if (!(filehandle = fopen(filename,"rb"))){
  printf("cannot open %s\n",filename);
  exit(1);     
 }
 return(filehandle);
}



ctronome.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include "ctronome.h"
#include "routines.h"

dword bytes_read;

int dsp_device;

dword dsp_pattern_length;

/* the two wav files */
byte wav1[1000000];
byte wav2[1000000];

dword wav_bytes_to_read;

byte wav_header[100];

const char *wave1, *wave2;

FILE *wavfile, *program;

byte is_program;
dword bpm_base_length;

dword c1,c2,c3,c4,lo;

word wav_channels;
dword wav_samplerate;
word wav_bitspersample;
word wav2_channels;
dword wav2_samplerate;
word wav2_bitspersample;

int bpt_base_specified,bpm_base_specified;
int i1,i2,i3,i4,i5;
byte tmp[8219];

int main(int argc,char *argv[]){
 parm_init(argc,argv);

 /* create an endless/pcount times loop */
 while(pcount > 0){
  if (is_program){
   next_program(program);
  }

  if (pcount > 0){
   if (is_program) printf("count: %d, bpm: %d/%d, bpt: %d/%d\n",count,bpm[0],bpm[1],bpt[0],bpt[1]);

   /* lets calculate the appropriate pattern length for our bpm and bpt */
   bpm_base_length = dsp_speed * dsp_depth * dsp_channels * 60 / bpm[0];
   dsp_pattern_length = bpm_base_length * bpm[1] / bpt[1];
   /* if (debug){
    printf("dsp_speed: >%d<, dsp_depth: >%d<, dsp_channels: >%d<, patt. length: >%d<\n",dsp_speed,dsp_depth,dsp_channels,
                                                                                        dsp_pattern_length);
   } */

   while(i3 = dsp_pattern_length % (dsp_depth * dsp_channels)){
    dsp_pattern_length++;
   }

   for (c4 = 0;c4 < count;c4++){
    dsp_write(dsp_device,wav1,dsp_pattern_length);
    for (c3 = bpt[0];c3 > 1; c3--){
     dsp_write(dsp_device,wav2,dsp_pattern_length);
    }
   }
   if (!(is_program)) pcount -= pdecrease;
  }
 }
 dsp_close;
 return(0);
}

void next_program(FILE *programfile){
 byte temp[8192];
 dword lo1;
 int lo2,lo3,lo4;
 /* a normal line should be at least 5 characters long + newline */
 lo2 = 0;
 while (lo2 >= 0){
  while ( (lo1 = getnextline(temp,programfile,8192)) < 6){
   /* if 0 character has been read, seek to the beginning of the file */
   if (lo1 < 1){
    fseek(programfile,SEEK_SET,0);
    pcount -= pdecrease;
    if (pcount) printf("repeat\n");
   }
  }
  lo2 = str_search(temp,hashmark);
 }
 count = atoi(temp);

 lo2 = str_search(temp,space);
 lo2++;
 lo2 += str_search(&temp[lo2],space);
 bpm_base_specified = bpt_base_specified = 0;

 if (lo2) {
  lo3 = str_search(&temp[++lo2],slash);
  bpt[0] = atoi(&temp[lo2]);

  if (lo3 >= 0){
   bpt[1] = atoi(&temp[lo2] + ++lo3);
   bpt_base_specified = 1;
  }
 }

 lo2 = str_search(temp,space);
 if (lo2) {
  lo3 = str_search(&temp[++lo2],slash);
  bpm[0] = atoi(&temp[lo2]);
  if (debug) printf("debug: prg: bpm0: '%d', lo3: '%d'\n",bpm[0],lo3);
  if (lo3 >= 0){
   bpm[1] = atoi(&temp[lo2] + ++lo3);
   bpm_base_specified = 1;
  }
 }

 /* some parameter post-processing */
 if (!(bpt_base_specified)){
  if (bpm_base_specified){
   bpt[1] = bpm[1];
  } else {
   bpt[1] = default_base_note;
  }
 }

 if (!(bpm_base_specified)){
  if (bpt_base_specified){
   bpm[1] = bpt[1];
  } else {
   bpm[1] = default_base_note;
  }
 }

 if (bpt[0] < 1) bpt[0] = 1; if (bpt[0] > 50) bpt[0] = 50;
 if (bpt[1] < 1) bpt[1] = 1; if (bpt[1] > 50) bpt[1] = 50;

 if (bpm[0] > 250) bpm[0] = 250; if (bpm[0] < 30) bpm[0] = 30;
 if (bpm[1] > 20) bpm[1] = 20; if (bpm[1] < 1) bpm[1] = 1;
}


void parm_init(int argc,char *argv[]){

 /* first, get the parameters */
 int i;
 dword bytes_read;

 debug = 0;
 bpm_base_specified = bpt_base_specified = 0;

 for (i = 1; i < argc; i++){
  /* debug */
  if ((strcmp(argv[i], "-debug") == 0) ||
      (strcmp(argv[i], "--debug") == 0)) {
   printf("debug mode\n");
   debug = 1;
  }
 }

 for (i = 1; i < argc; i++){
  /* help */
  if ((strcmp(argv[i], "-h") == 0) ||
      (strcmp(argv[i], "--h") == 0) ||
      (strcmp(argv[i], "-help") == 0) ||
      (strcmp(argv[i], "--help") == 0)) {
   printf(HELP);
   exit(0);
  }

  /* version and credits */
  if ((strcmp(argv[i], "-v") == 0) ||
      (strcmp(argv[i], "--v") == 0) ||
      (strcmp(argv[i], "-version") == 0) ||
      (strcmp(argv[i], "--version") == 0)) {
   printf("%s v%s\n%s",MYNAME,VERSION,CREDITS);
  }

  /* wav1 */
  if ((strcmp(argv[i], "-w1") == 0) && (i + 1 < argc)){
   metronomewav1 = argv[++i];
   if (debug) printf("debug: wav1: '%s'\n",metronomewav1);
  }

  /* wav2 */
  if ((strcmp(argv[i], "-w2") == 0) && (i + 1 < argc)){
   metronomewav2 = argv[++i];
   if (debug) printf("debug: wav2: '%s'\n",metronomewav2);
  }

  /* dsp device */
  if ((strcmp(argv[i], "-d") == 0) && (i + 1 < argc)){
   dspdev = argv[++i];
   if (debug) printf("debug: dsp: '%s'\n",dspdev);
  }

  /* bpt */
  if ( (strcmp(argv[i], "-t") == 0) && (i + 1 < argc)){
   i1 = str_search(argv[++i],slash);
   bpt[0] = atoi(argv[i]);
   if (i1 >= 0){
    bpt[1] = atoi(argv[i] + ++i1);
    bpt_base_specified = 1;
   }
   if (debug) printf("debug: bpt: '%d'/'%d'\n",bpt[0],bpt[1]);
  }

  /* bpm */
  if ((strcmp(argv[i], "-b") == 0) && (i + 1 < argc)){
   i2 = str_search(argv[++i],slash);
   bpm[0] = atoi(argv[i]);
   if (i2 >= 0){
    bpm[1] = atoi(argv[i] + ++i2);
    bpm_base_specified = 1;
   }
   if (debug) printf("debug: bpm: '%d'/'%d'\n",bpm[0],bpm[1]);
  }

  /* pcount */
  if ((strcmp(argv[i], "-c") == 0) && (i + 1 < argc)){
   pcount = atoi(argv[++i]);
   pdecrease = 1;
   if (debug) printf("debug: count: '%s'\n",pcount);
  }

  /* program file */
  if ((strcmp(argv[i], "-p") == 0) && (i + 1 < argc)){
   programfile = argv[++i];
   is_program = 1;
   if (debug) printf("debug: program: '%s'\n",programfile);
  }
 }

 /* some parameter post-processing */
 if (!(bpt_base_specified)){
  if (bpm_base_specified){
   bpt[1] = bpm[1];
  } else {
   bpt[1] = default_base_note;
  }
 }

 if (!(bpm_base_specified)){
  if (bpt_base_specified){
   bpm[1] = bpt[1];
  } else {
   bpm[1] = default_base_note;
  }
 }

 if (bpm[0] > 250) bpm[0] = 250; if (bpm[0] < 30) bpm[0] = 30;
 if (bpm[1] > 20) bpm[1] = 20; if (bpm[1] < 1) bpm[1] = 1;

 if (bpt[0] < 1) bpt[0] = 1; if (bpt[0] > 50) bpt[0] = 50;
 if (bpt[1] < 1) bpt[1] = 1; if (bpt[1] > 50) bpt[1] = 50;

 /* cleanup buffers */
 for (c1 = 0; c1 < 1000000; c1++){
  wav1[c1] = 0;
  wav2[c1] = 0;
 }

 /* open wav file 1 */
 wavfile = openread(metronomewav1);

 /* read the header first */
 bytes_read = fread(&wav_header,1,44,wavfile);
 if (bytes_read < 44){
  printf("wav file %s too short\n",metronomewav1);
  exit(1);
 }

 wav_channels = * (word *) &wav_header[22];
 wav_samplerate = * (dword *) &wav_header[24];
 wav_bitspersample = * (word *) &wav_header[34];

 if (debug)
  printf("debug: wav channels: '%d', samplerate: '%d', bits per sample: '%d'\n",wav_channels,wav_samplerate,wav_bitspersample);

 if (debug) printf("debug: calling dsp_init(%s)\n",dspdev);

 dsp_device = dsp_init(dspdev,wav_bitspersample,wav_channels,wav_samplerate);

 if (debug)
  printf("debug: dsp channels: '%d', samplerate: '%d', bits per sample: '%d'\n",dsp_channels,dsp_speed,dsp_format);

 dsp_depth = dsp_format / 8;

 wav_bytes_to_read = dsp_depth * dsp_channels * dsp_speed / 2;
 if (debug) printf("debug: maximum wav bytes to read: '%d'\n",wav_bytes_to_read);

 bytes_read = fread(&wav1,1,wav_bytes_to_read,wavfile);
 if (debug) printf("debug: wav1 bytes read: '%d'\n",bytes_read);
 if (bytes_read < 10){
  printf("wav file %s too short\n",metronomewav1);
  exit(1);
 }

 fclose(wavfile);

 /* open wav file 2 */
 wavfile = openread(metronomewav2);

 /* read the header first */
 bytes_read = fread(&wav_header,1,44,wavfile);
 if (bytes_read < 44){
  printf("wav file %s too short\n",metronomewav2);
  exit(1);
 }

 wav2_channels = * (word *) &wav_header[22];
 wav2_samplerate = * (dword *) &wav_header[24];
 wav2_bitspersample = * (word *) &wav_header[34];

 if (debug)
  printf("debug: wav2 channels: '%d', samplerate: '%d', bits per sample: '%d'\n",wav2_channels,wav2_samplerate,wav2_bitspersample);

 if ((wav_channels != wav2_channels) ||
     (wav_bitspersample != wav2_bitspersample)){
  printf("the two WAV files must have the same number of channels and same samplerate\n");
  exit(1);
 }

 if ((debug) && (wav_samplerate != wav2_samplerate)){
  printf("debug: wav1 and wav2 samplerate differs, may sound funny\n");
 }

 /* skip the wav header */
 /* fseek(wavfile,44,SEEK_SET); */

 bytes_read = fread(&wav2,1,wav_bytes_to_read,wavfile);
 if (debug) printf("debug: wav2 bytes read: '%d'\n",bytes_read);
 if (bytes_read < 10){
  printf("wav file %s too short\n",metronomewav2);
  exit(1);
 }
 fclose(wavfile);

 /* open wav file 2 */
 if (is_program) program = openread(programfile);

 printf ("bpm: %d/%d, bpt: %d/%d",bpm[0],bpm[1],bpt[0],bpt[1]);
 if (pdecrease) printf ("repeat count: %d",pcount);
 if (is_program) printf ("\nprogram file: %s",programfile);
 printf ("\n");
}

